Sort on other features when first comparison matches

eslindsey opened the pull request:
Should fix #214 ... bot was only sorting first thing asked for (IV or CP), and if they matched it did not look at anything else (I guess it picked the first one in the list).

I modified the following behaviors:
`GetHighestPokemonOfTypeByCp`
Old behavior: Sort descending by CP
New behavior: Sort descending by CP, then descending by IV
`GetHighestPokemonOfTypeByIv`
Old behavior: Sort descending by IV
New behavior: Sort descending by IV, then descending by CP
`GetHighestsCp`
Old behavior: Sort descending by CP, then increasing by StaminaMax (why increasing?)
New behavior: Sort descending by CP, then descending by IV
 `GetHighestsPerfect`
Old behavior: Sort descending by IV
New behavior: Sort descending by IV, then descending by CP

As an example: If our subjects are 100 IV, 1250 CP vs. IV 100 IV, 900 CP, the lack of a second comparison would sometimes result in the bot discarding the 1250 CP Pokemon if it was added to the list after the 900 CP one.

eslindsey committed:
Sort on other features when first comparison matches

Fenixxia commented:
The second parameter needs to be LV. Not "first in list". Sudden case again possible similar errors.

DurtyFree commented:
Is this tested to be working @eslindsey ?

eslindsey commented:
When asking for highest IV, I added CP as the second criteria. When asking for highest CP, I added IV as the second criteria. @Fenixxia, please explain why LV should be used, when it is not used in any other logical considerations throughout the bot.

kyfranky commented:
@eslindsey  maybe because  this sir #4 , Fenixxia suggestion

eslindsey commented:
LV (if that means level, it very difficult to tell) is represented by CP well enough (higher level Pokemon have higher CP). LV is much more likely to collide than CP, so it makes sense to use CP. I am asking for a clear explanation indicating otherwise, if that is to be the case.

marksterxxx commented:
This will barely work because, difference between perfection and cp/lv is very discrete/small. For example, IV then CP, If your perfection is 90.5, the next pokemon should also have 90.5 for cp sorting to kick in, which is highly unlikely.

ShawnHardwick commented:
@marksterxxx I disagree. For people who have 100IV Pokemon who have dumped millions of stardust into, I don't want my bot to catch another same 100IV Pokemon and then discard my high CP/IV pokemon. 

You are right though. There is no way to weight how much offset should be accounted for when determining which takes priority. Would you rather have a 95IV 600CP Pikachu or a 98IV 50CP Pikachu? Thats for the owner to decide, and while this doesn't solve that issue, it is a step in the right direction.

Fenixxia commented:
@eslindsey In this and in other programs began to use the CP as the primary parameter is only because of ignorance of the basics of the game a lot of people who play this game, as well as developers who listen to the queries of these people. Although information about the internal structure and hidden parameters, their importance appeared quite early, but was not massively distributed. The majority thought that the more CP the better. 
When IV and was understood more deeply, the programs appeared and this option. But to move away from the CP still can't. CP is just results of a specific formula, it is a common performance indicator pokemon. And quite subjective. CP includes the parameters from the attack and the level of x and protection... because these different parameters can be different levels and different importance level, we obtain two pokemon with the same CP, but they will have very different parameters, better or worse than each other in the Arena. 
By this I insist that sort of pokemon were made on the critical indicators. Have a clear view what should be the perfect pokemon-candidate for the optimal and economical for its development in the future, without wasting dust, time, sweets and energy.
1) IV 100% - perspektivet pokemon (does not change during the evolution of a stable parameter)
2) Have a STAB - presence gives a bonus to attack 1.25 (may vary for different type of attack with the evolution of the unstable parameter)
3) LV tends to 30 - the higher the level the less sweets and to dust we will spend on further upgrades (big savings). Pokemon above level 30 in the game to catch it is impossible.
4) the final form of the transformation (we guarantee the preservation of the type of attack) at each stage of the transformation type attacks of the pokemon may change randomly. Pokemon without STAB will lose in the attack pokemon with STAB, even if the second IV will be less as the level.

I apologize for the speech patterns. My native language is quite difficult to translate.

Fenixxia commented:
I'm trying to explain just what way of development (search for pokemon) robot program is now suboptimal, flawed result. And You discuss how this flawed to fix, but in fact it will be flawed again but differently. 
У Вас сейчас есть шанс поменять направление пути. Стать программой роботом верного направления. К тому моменту когда массовая публика поймет какие параметры действительно важны у вас уже будет нужный инструмент. Инструмент, который действительно приносит верный результат прямым путем, а не приближенный к верному окольным.

UntilSunrise commented:
@DurtyFree : 
Would prefer @Fenixxia solution

kyfranky commented:
we have move ranking now #299

marksterxxx commented:
My idea is to change PrioritizeIVOverCp to PrioritizeIVOverCpPercentage,
ex: PrioritizeIVOverCpPercentage = 60
Then get CpPercent = CurrentCp/PoweredCp    , where PoweredCp is the Cp @ max stardust
We can then have BattleRating = (Perfection)*60% + CpPercent*(100-60)%.
Those who wish pure IV can set PrioritizeIVOverCPPercentage to 100, while the economic ones who want mass quantity can add weight to Cp
![untitled](https://cloud.githubusercontent.com/assets/2887363/17318255/78160c0e-58b7-11e6-811f-38359b46268d.png)

kyfranky commented:
@marksterxxx sorry, but what is formula for PoweredCp ?  as reference from [https://thesilphroad.com/research](url) 100% snorlax will have 3112 CP ?

eslindsey commented:
@UntilSunrise what "solution" has been provided by @Fenixxia?

This is a simple fix, that addresses a simple problem: if you have two 100 IV Pokemon, it should prefer the one with higher CP. Right now it doesn't, and it will sometimes discard a Pokemon which is better in *all* respects, versus a lower level one.

Nimryel commented:
Lots of discussion on unrelated matters. This merely adds secondary ordering on CP (IV is still primary) so that the results go
989 CP | 100 IV
1300 CP  | 100 IV
1400 CP | 100 IV
Instead of randomly ie.
1400 CP 100 IV
939 CP  100 IV
1300 CP 100 IV
It is expected behavior, should have been merged ages ago.

mvegter197 commented:
+1 on merging

Afraidium commented:
+1, this solves the problem at hand, in a logical and beneficial way. For the concerns of others, we need to add my filtering before the final shot! BANG!

kyfranky commented:
@eslindsey sir if u can add 1 more ordering related move set rank from this  #4, it will more awesome sir.

eslindsey commented:
@kyfranky the likelihood of IV *and* CP matching is very low, so I don't know that adding a third comparison is necessary. The reason I selected CP as the second sorting behavior is because it will satisfy most of the user base...it is a good *overall* indication of "how good" a Pokemon is.

kyfranky commented:
@eslindsey okay sir, thankyou

eslindsey commented:
@DurtyFree this is tested and working for me.

marksterxxx commented:
@kyfranky  forgive me for the off topic post, but to answer your question, poweredCP is the cp when you max stardust the pokemon cp at your level ,  pokemon max level is 1.5 + trainer level, compute pokemon stats for that level using base stats and cp multiplier and you can compute the cp of that pokemon for your level.

eslindsey commented:
@marksterxxx That would still confuse the heck out of people that see higher *current* CP Pokemon getting transferred over lower ones...I understand the benefit of performing calculations that way, but it sounds like that should be reserved as an advanced feature.

UntilSunrise commented:
i agree that this is a hotfix, but we should redefine the potential power calculation like Fenixxia mentioned.

Fenixxia commented:
I should add, if we have the selection of pokemon was my way, the question which we raised at first in General would not have happened. we would not get up the choice on what pokemon to remove at 100% IV. the following selection criteria my method would solve this logic in the primary selection, not the final choice for deletion. You are treating the consequences. My method allows for these effects to have nothing at all.
